File: Data Structures & Algorithms/invert-a-binary-tree/submission-3.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {TreeNode}
     */
    invertTree(root: TreeNode | null): TreeNode {
        // Base case: if the node is null, stop recursion
        if (root === null) {
            return null;
        }

        // Swap using array destructuring (matches Python syntax)
        [root.left, root.right] = [root.right, root.left];

        // Recursively call on the subtrees
        this.invertTree(root.left);
        this.invertTree(root.right);

        return root;
    }
}

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-3.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    private int sol = int.MinValue;

    public int MaxPathSum(TreeNode root) {
        int _ = Preorder(root);
        return sol;
    }

    private int Preorder(TreeNode node) {
        if (node is null) {
            return 0;
        }

        int left_sum = Math.Max(Preorder(node.left), 0);
        int right_sum = Math.Max(Preorder(node.right), 0);

        int curr_path = node.val + left_sum + right_sum;
        sol = Math.Max(sol, curr_path);

        return node.val + Math.Max(left_sum, right_sum);
    }
}
